Add a guide to PWA caching strategies (#27442)
mdn/content: files/en-us/web/progressive_web_apps/guides/caching/index.md

### Cache first

In this strategy we will precache some resources, and then implement a "cache first" strategy only for those resources. That is:

- For the precached resources, we will:
  - Look in the cache for the resource, and return the resource if it is found.
  - Otherwise, go to the network. If the network request succeeds, cache the resource for next time.
- For all other resources, we will always go to the network.

Precaching is an appropriate stategy for resources that the PWA is certain to need, that will not change for this version of the app, and that need to be fetched as quickly as possible. That includes, for example, the basic user interface of the app. If this is precached, then the app's UI can be rendered on launch without needing any network requests.

First, the service worker precaches static resources in its `install` event handler:

```js
const cacheName = "MyCache_1";
const precachedResources = ["/", "/app.js", "/style.css"];

async function precache() {
  const cache = await caches.open(cacheName);
  return cache.addAll(precachedResources);
}

self.addEventListener("install", (event) => {
  event.waitUntil(precache());
});
```

> In the `install` event handler, we pass the result of the caching operation into the event's {{domxref("ExtendableEvent.waitUntil", "waitUntil()")}} method. This means that if caching fails for any reason, the installation fails: conversely, if installation succeeded, the service worker can be sure that the resource was added to the cache.

Reviewer: ```suggestion
In the `install` event handler, we pass the result of the caching operation into the event's {{domxref("ExtendableEvent.waitUntil", "waitUntil()")}} method. This means that if caching fails for any reason, the installation of the service worker fails: conversely, if installation succeeded, the service worker can be sure that the resource was added to the cache.
```

(to make sure readers don't mistake this for the installation of the PWA).

    }
    return networkResponse;
  } catch (error) {
    return Response.error();
  }
}

self.addEventListener("fetch", (event) => {
  if (precachedResources.includes(url.pathname)) {
    event.respondWith(cacheFirst(event.request));
  }
});
```

We return the resource by calling the event's {{domxref("FetchEvent.respondWith()", "respondWith()")}} method. If we don't call `respondWith()` for a given request, then the request is sent to the network as if the service worker had not intercepted it. So if a request is not precached, it just goes to the network.

When we add `networkResponse` to the cache, we must clone the response and add the copy to the cache, returning the original. This is because `Response` objects are streamable, so may only be read once.

You might wonder why we fall back to the network for precached resources. If they are precached, can't we be sure they will be in the cache? The reason is that it is possible for the cache to be cleared, either by the browser or by the user. Although this is unlikely, it would make the PWA unusable unless it can fall back to the network. See [Deleting cached data](#deleting_cached_data).

### Cache first with cache refresh

> The drawback of "cache first" is that once a response is in the cache, it is never refreshed until a new version of the app is installed.

Reviewer: ```suggestion
The drawback of "cache first" is that once a response is in the cache, it is never refreshed until a new version of the service worker is installed.
```

Sure, new versions of a PWA require a new version of the SW to be installed. But it's possible to just change 1 bit in the SW and the browser will go and fetch it, and install it again. So, I think it's more correct to just talk about the SW here, not the entire app.

### Network first

The last strategy we'll look at, "network first", is the inverse of cache first: we try to retrieve the resource from the network. If the network request succeeds, we return the response and update the cache. If it fails, we try the cache.

This is useful for requests for which it is important to get the most fresh response possible, but for which a cached resource is better than nothing. A messaging app's list of recent messages might fall into this category.

```js
function isCachable(request) {
  const url = new URL(request.url);
  return !url.pathname.endsWith(".json");
}

async function networkFirst(request) {
  try {
    const networkResponse = await fetch(request);
    if (networkResponse.ok) {
      const cache = await caches.open("MyCache_1");
      cache.put(request, networkResponse.clone());
    }
    return networkResponse;
  } catch (error) {
    const cachedResponse = await caches.match(request);
    return cachedResponse || Response.error();
  }
}

self.addEventListener("fetch", (event) => {

>   if (isCachable(event.request)) {

Reviewer: To get the point across even better, we might want to rename isCachable to something else. Maybe isMessageListRequest.
And then in the function's body, change the test to something else. Maybe something that checks that the URL matches some message API name or something.
This would better match your description above the code snippet.

Author: I've done this slightly differently in https://github.com/mdn/content/pull/27442/commits/dfb782de5464383da286df38728f8a4afd1cdbdd.

This means that whether and when you should cache a resource is very dependent on the resource in question, and a PWA will typically adopt different strategies for different resources. In this guide we'll look at some common caching strategies for PWAs, and see which strategies make sense for which resources.

## Caching technology overview

The main technologies on which a PWA can build a caching strategy are the [Fetch API](/en-US/docs/Web/API/Fetch_API), the [Service Worker API](/en-US/docs/Web/API/Service_Worker_API), and the [Cache API](/en-US/docs/Web/API/Cache).

### Fetch API

The Fetch API defines a global function {{domxref("fetch()")}} for fetching a network resource, and {{domxref("Request")}} and {{domxref("Response")}} interfaces that represent network requests and responses. The `fetch()` function takes a `Request` or a URL as an argument, and returns a {{jsxref("Promise")}} that resolves to a `Response`.

The `fetch()` function is available to service workers as well as to the main app thread.

### Service Worker API

A service worker is a part of a PWA: it's a separate script that runs in its own thread, separate from the app's main thread.

> When a service worker is installed, the browser fires an event called {{domxref("ServiceWorkerGlobalScope.install_event", "install")}} in the service worker's global scope. At this point the service worker can _precache_ resources, fetching them from the network and storing them in the cache. This ensures that they will already be available locally the first time they are needed.

Author: We should probably add a note that installing the SW is different from installing the PWA. The SW can be installed without the user ever installing your app as a PWA.

Reviewer: Agreed!

Author: Oh, also added in https://github.com/mdn/content/pull/27442/commits/7df4572a0baa322a52372eba7cab4977538f098d.

### Cache API

The {{domxref("Cache")}} interface provides persistent storage for `Request`/`Response` pairs. It provides methods to add and delete `Request`/`Response` pairs, and to look up a cached `Response` matching a given `Request`. The cache is available in both the main app thread and the service worker: so it is possible for one thread to add a response there, and the other to retrieve it.

Most commonly, the service worker will add resources to the cache in its `install` or `fetch` event handlers.

## Caching strategies

A caching strategy is an algorithm for when to cache a resource, when to serve a cached resource, and when to get the resource from the network. In this section we'll summarize some common strategies.

This isn't an exhaustive list: it's just intended to illustrate the kinds of approaches a PWA can take. Each PWA must consider its own specific needs when deciding which strategies to use for which resources.

### Cache only

> In this strategy, the service worker precaches some resources in its `install` event handler:

Reviewer 2: It might be just me, but I interpreted "Cache only" initially as a strategy for _the whole PWA_. What you actually are describing is a strategy for **some** resources in a PWA that might implement multiple strategies. I was therefore going to "correct" this line to "most resources".

Worth pointing out?

Author: In https://github.com/mdn/content/pull/27442/commits/cac85eb2e1b8a78b0ba488c64d757b7dbd6f7b67 I added a para at the start of this section that hopefully clarifies things.

```js
const cacheName = "MyCache_1";
const precachedResources = ["/", "/app.js", "/style.css"];

async function precache() {
  const cache = await caches.open(cacheName);

>   return cache.addAll(precachedResources);

Reviewer 2: It is pretty cool that you can just tell the cache to do the fetching - these being either requests or URLs. Might be worth noting that, since I was looking for some kind of fetch operation and then addition.

Reviewer 2: Note also this returns a promise, which makes sense because you later wait on the event to complete. However there is quite a lot going on here you might want to talk about. Certainly consider linking to https://developer.mozilla.org/en-US/docs/Web/API/ExtendableEvent/waitUntil since this is not a "normal" event and that has some useful info in it.

Reviewer 2: Ignore, reading top down.

}

self.addEventListener("install", (event) => {
  event.waitUntil(precache());
});

> ```

Author: Here's a thing I don't understand. Even if you precache as an installation dependency, it's possible (although unlikely) that the cache might be cleared (either by the user or by the browser) under your feet. If that happens, and you have cache-only, is this catastrophic for your app? If so, would it be better to omit this one and start with cache-first?

Reviewer: Realistically, I don't know that many PWAs will want a cache-only strategy. In the end, the user is in charge, and they can delete the cache whenever they want. So if your app never tries the network ever again, the app will be broken beyond repair, unless you ctrl+shift+R to force reload from the network (or use devtools to unregister the SW).

To me pre-cache is just a first step for other strategies (cache-first). First, you put your assets in cache, and then you intercept requests to these assets in the fetch handler, and respond with the cached version. If, for any reason, this doesn't work (perhaps because the user emptied the cache), then you go to the network again and then cache.

Author: >  So if your app never tries the network ever again, the app will be broken beyond repair, unless you ctrl+shift+R to force reload from the network (or use devtools to unregister the SW).

Yeah, that's exactly what I was thinking. So showing cache-only is perhaps going to mislead people, and we should start with cache-first.

Reviewer 2: Yes ^^^^. 

I would rename this strategy to Precaching. If you make it clear that this is part of a strategy as per https://github.com/mdn/content/pull/27442/files#r1242991321 it might then require very little modification.

Author: OK I have pushed https://github.com/mdn/content/pull/27442/commits/7df4572a0baa322a52372eba7cab4977538f098d, which does 2 related things:
- add a new section on "when to cache" that talks about precaching as well as other times you might cache. I kept it fairly short though and didn't include code here as I want to keep the code together in the "caching strategies" section.
- retired "cache only" and basically folded the precaching stuff into "cache first"

> Note that in the `install` event handler, we pass the result of the caching operation into the event's {{domxref("ExtendableEvent.waitUntil", "waitUntil()")}} method. This means that if caching fails for any reason, the installation fails: so conversely, when the service worker intercepts a request for a precached resource, it can be assured that the resource exists in the cache.

Reviewer 2: "it can be assured that the resource exists in the cache." - unless the cache is later deleted.

The wording around "so conversely" is a little hard to parse - I think something like "conversely, if the installation succeeds, ...."

Author: -> https://github.com/mdn/content/pull/27442/commits/b49133322674428d48fb2eb9faba3fdb62c81343.

  const url = new URL(event.request.url);
  if (precachedResources.includes(url.pathname)) {
    event.respondWith(cacheOnly(event.request));
  }
});
```

This is an appropriate stategy for resources that the PWA is certain to need, that will not change for this version of the app, and that need to be fetched as quickly as possible. That includes, for example, the basic user interface of the app. If this is precached, then the app's UI can be rendered on launch without needing any network requests.

Note that in the example above, `respondWith()` is only called for requests in `precachedResources`. If `respondWith()` is not called for a given request, the request is sent to the network as if the service worker had not intercepted it. So this strategy actually implements "cache only" for precached resources, and "network only" for all other requests.

### Cache first

This is an extension to the previous strategy. In the `fetch` event handler, if the resource is cachable, we:

- Look in the cache for the resource, and return the resource if it is found.
- Otherwise, go to the network. If the network request succeeds, cache the resource for next time.

> If the resource is not cachable, we only use the network, as before. The definition of "cachable" is specific to the PWA, but again includes the idea that the resource should not change for this version of the app. In this example, everything except requests for JSON data is treated as cachable.

Reviewer 2: ```suggestion
If the resource is not cachable, we only use the network, as before. The definition of "cachable" is specific to the PWA, but again includes the idea that the resource should not change for this version of the app. In this example, everything except requests for JSON data are treated as cachable.
```

```js

> function isCachable(request) {

Author: I think this is redundant now.

  const url = new URL(request.url);

>   return !url.pathname.endsWith(".json");

Reviewer: Maybe add a comment here saying that, only for this example, we only consider non .json requests as cachable.
Again, for the sake of people who just copy/paste code without reading the docs before it :)

}

async function cacheFirst(request) {
  const cachedResponse = await caches.match(request);
  if (cachedResponse) {
    return cachedResponse;
  }
  try {
    const networkResponse = await fetch(request);
    if (networkResponse.ok) {
      const cache = await caches.open("MyCache_1");
      cache.put(request, networkResponse.clone());
    }
    return networkResponse;
  } catch (error) {
    return Response.error();
  }
}

self.addEventListener("fetch", (event) => {
  if (isCachable(event.request)) {
    event.respondWith(cacheFirst(event.request));
  }
});
```

Note also that when we add `networkResponse` to the cache, we must clone the response and add the copy to the cache, returning the original. This is because `Response` objects are streamable, so may only be read once.

> ### Stale while revalidate

Reviewer 2: Is this a well understood name for the concept? It doesn't feel like it tells me what this does.  If not, perhaps something like "Cache first with cache refresh"

Author: Yes, this is unfortunately what everyone calls it: https://developer.chrome.com/docs/workbox/caching-strategies-overview/#stale-while-revalidate.

Reviewer 2: Given how awful that name is I would rename to  `Cache first with refresh` and in the body have "this is also referred to as stale while revalidate"  in order to help search engines.

Your call though. I like the caching topic too.

Author: -> https://github.com/mdn/content/pull/27442/commits/64c618d0fe3400f0c9bbc5c20d98c331579e38ab

The drawback of "cache first" is that once a response is in the cache, it is never refreshed until a new version of the app is installed.

The "stale while revalidate" strategy is similar to "cache first", except that we always send the request to the network, even after a cache hit, and use the response to refresh the cache. This means we get the responsiveness of "cache first", but get a fairly fresh response (as long as the request is made reasonably often).

```js
function isCachable(request) {
  const url = new URL(request.url);
  return !url.pathname.endsWith(".json");
}

async function staleWhileRevalidate(request) {
  const cachedResponse = await caches.match(request);
  const fetchedResponse = fetch(request);

  fetchedResponse.then(async (networkResponse) => {
    if (networkResponse.ok) {
      const cache = await caches.open("MyCache_1");
      cache.put(request, networkResponse.clone());
    }
    return networkResponse;
  });

>   return cachedResponse || fetchedResponse;

Author: I'm a bit confused about the async stuff here.

Part of the point of this pattern AIUI is that you can return the cached value before the network request has succeeded/failed, isn't it?

But you have to clone the response before you return it, because you can't clone a response that's been consumed. I'm a bit confused about the async stuff here. Part of the point of this pattern AIUI is that you can return the cached value before the network request has succeeded/failed, isn't it?

Is there a better way to express this using `await`?

Reviewer 2: > Part of the point of this pattern AIUI is that you can return the cached value before the network request has succeeded/failed, isn't it?

FWIW This was my concern too, and I don't think this is fixed by the suggestion above ^^^. I also wonder if for some reason the cache.match took forever you'd have to wait on that. 

Could you shove both the promise into [Promise.any()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any) and return the result of the promise as soon as it arrives?

Author: OK, I updated the implementation along the lines suggested by Patrick and done some testing, and it seems to work :) -> https://github.com/mdn/content/pull/27442/commits/4e9ee84cab6d1844fc4d4ce132f5c5c8dd5c1fe2.

>> Part of the point of this pattern AIUI is that you can return the cached value before the network request has succeeded/failed, isn't it?
>
> FWIW This was my concern too, and I don't think this is fixed by the suggestion above ^^^.

I think it is addressed (in both versions), because if we get a cache hit we return the cached response without waiting for `fetchResponsePromise`. I experimented with a version of this that had some logging!

```js
// main.js
fetchIngredients.addEventListener("click", async () => {
  try {
    const response = await fetch("ingredients.json");
    const ingredients = await response.json();
    console.log(`app received response to ingredients.json`);
  } catch (e) {
    console.log("error:");
    console.log(e);
  }
});
```

```js
// sw.js
async function cacheFirstWithRefresh(request) {
  const fetchResponsePromise = fetch(request).then(async (networkResponse) => {
    console.log(`sw received network response to ${request.url}`);
    if (networkResponse.ok) {
      const cache = await caches.open("MyCache_1");
      cache.put(request, networkResponse.clone());
    }
    return networkResponse;
  });

  const cached = await caches.match(request);
  if (cached) {
    console.log(`sw: from cache: ${request.url}`);
    return cached;
  } else {
    console.log(`sw: from network: ${request.url}`);
    return fetchResponsePromise;
  }
}

self.addEventListener("fetch", (event) => {
  event.respondWith(cacheFirstWithRefresh(event.request));
});
```

So the first time I press the button, there is no resource in the cache (it's not precached) and in the console I get:

<img width="500" alt="Screen Shot 2023-06-27 at 9 10 13 PM" src="https://github.com/mdn/content/assets/432915/b7f04d06-a8a2-4fbb-affe-e9bbdbba6359">

This makes sense:

- `sw: from network: https://localhost:4443/ingredients.json`: we have to return the network resource, because the cache is empty
- `sw received network response to https://localhost:4443/ingredients.json`: the app has to wait till the response is received before it can get the JSON. At the point the SW can clone and cache it
- `app received response to ingredients.json`: the app finally gets the JSON

Next time, the response is in the cache, so the SW sends back the cached version and fetches a fresh one:

<img width="500" alt="Screen Shot 2023-06-27 at 9 11 44 PM" src="https://github.com/mdn/content/assets/432915/f9057412-dded-4781-b703-61477bb34cd9">

...so this time, the app gets the JSON before the response has been received.

At least that's what I think is going on! I'm not sure if this is very reliable way to test it though.

> I also wonder if for some reason the cache.match took forever you'd have to wait on that.

Yes, it's an assumption that cache access is usually much faster than network access. There are caching strategies that are supposed to deal with slow caches, but they seemed a little niche to me.

> Could you shove both the promise into [Promise.any()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any) and return the result of the first promise as soon as it fulfills?

Not quite, because [`match()`](https://developer.mozilla.org/en-US/docs/Web/API/Cache/match) doesn't reject on a cache miss, it returns `undefined` (and in that case, you want to return the network response).

I suppose you could just return the network version if the cache call hasn't resolved. I'm not sure it's worth the complexity though...

This is a good choice when responsiveness is important, and freshness is somewhat important but not essential.

### Network first

The last strategy we'll look at, "network first", is the inverse of cache first: we try to retrieve the resource from the network. If the network request succeeds, we return the response and update the cache. If it fails, we try the cache.

```js
function isCachable(request) {
  const url = new URL(request.url);
  return !url.pathname.endsWith(".json");
}

async function networkFirst(request) {
  try {
    const networkResponse = await fetch(request);
    if (networkResponse.ok) {
      const cache = await caches.open("MyCache_1");
      cache.put(request, networkResponse.clone());
    }
    return networkResponse;
  } catch (error) {
    const cachedResponse = await caches.match(request);
    return cachedResponse || Response.error();
  }
}

self.addEventListener("fetch", (event) => {
  if (isCachable(event.request)) {
    event.respondWith(networkFirst(event.request));
  }
});
```

> This is useful for requests for which it is important to get the most fresh response possible, but for which a cached resource is better than nothing. A messaging app's list of recent messages might fall into this category.

Reviewer 2: I would move this up above the code with the intro - just a preference to state the value of a strategy before explaining how it works.

Author: -> https://github.com/mdn/content/pull/27442/commits/bdb4c6c4905eb3e201716a537926f614c7345925

There are still requests for which no response is better than a possibly-outdated response, and for which only a "network only" strategy is appropriate. For example, if an app is showing the list of available products, it will be frustrating to users if the list is out of date.

## Deleting cached data

Caches have a limited amount of storage space, and the browser may evict an app's cached data if the limit is exceeded. The specific limits and behavior are browser-specific: see [Storage quotas and eviction criteria](/en-US/docs/Web/API/Storage_API/Storage_quotas_and_eviction_criteria) for details. In practice, eviction of cached data is a very rare event. The user can also clear an app's cache at any time.

> A PWA should clean up any old versions of its cache in the service worker's {{domxref("ServiceWorkerGlobalScope.activate_event", "activate")}} event: when this event fires, the service worker can be sure that no previous versions of the service worker are running, so old cached data is no longer needed.

Reviewer: Since all other sections have code samples, maybe add a code sample here too?
